californium-core/src/main/java/org/eclipse/californium/core/network/CorrelationContextMatcherFactory.java

```diff
@@ -13,30 +13,55 @@
+ *    Achim Kraus (Bosch Software Innovations GmbH) - create CorrelationContextMatcher
+ *                                      related to connector
  ******************************************************************************/
 package org.eclipse.californium.core.network;
 
+import org.eclipse.californium.core.coap.CoAP;
 import org.eclipse.californium.core.network.config.NetworkConfig;
+import org.eclipse.californium.elements.Connector;
 import org.eclipse.californium.elements.CorrelationContextMatcher;
-import org.eclipse.californium.elements.RelaxedCorrelationContextMatcher;
-import org.eclipse.californium.elements.StrictCorrelationContextMatcher;
+import org.eclipse.californium.elements.RelaxedDtlsCorrelationContextMatcher;
+import org.eclipse.californium.elements.StrictDtlsCorrelationContextMatcher;
+import org.eclipse.californium.elements.UdpCorrelationContextMatcher;
 
 /**
  * Factory for correlation context matcher.
  */
 public class CorrelationContextMatcherFactory {
 
 	/**
-	 * Create correlation context matcher according the configuration. If
+	 * Create correlation context matcher related to connector according the
+	 * configuration. If connector supports "coaps:" and
 	 * USE_STRICT_RESPONSE_MATCHING is set, use
-	 * {@link StrictCorrelationContextMatcher}, otherwise
-	 * {@link RelaxedCorrelationContextMatcher}.
+	 * {@link StrictDtlsCorrelationContextMatcher}, otherwise
+	 * {@link RelaxedDtlsCorrelationContextMatcher}. For other protocol flavors
+	 * the corresponding matcher is used. Note: currently the TCP based
+	 * correlation context matcher are still missing and therefore for backwards
+	 * compatibility the DTLS ones are used.
 	 * 
+	 * @param connector connector to create related correlation context matcher.
 	 * @param config configuration.
 	 * @return correlation context matcher
 	 */
-	public static CorrelationContextMatcher create(NetworkConfig config) {
-		return config.getBoolean(NetworkConfig.Keys.USE_STRICT_RESPONSE_MATCHING) ? new StrictCorrelationContextMatcher()
-				: new RelaxedCorrelationContextMatcher();
+	public static CorrelationContextMatcher create(Connector connector, NetworkConfig config) {
+		if (null != connector) {
+			if (connector.isSchemeSupported(CoAP.COAP_URI_SCHEME)) {
+				return new UdpCorrelationContextMatcher();
+			} else if (connector.isSchemeSupported(CoAP.COAP_SECURE_TCP_URI_SCHEME)) {
+				/*
+				 * To be implemented in a future PR, in the meanwhile use
+				 * default dtls matcher as default for backwards compatibility
+				 */
+			} else if (connector.isSchemeSupported(CoAP.COAP_TCP_URI_SCHEME)) {
+				/*
+				 * To be implemented in a future PR, in the meanwhile use
+				 * default dtls matcher as default for backwards compatibility
+				 */
+			}
+		}
+		return config.getBoolean(NetworkConfig.Keys.USE_STRICT_RESPONSE_MATCHING) ? new StrictDtlsCorrelationContextMatcher()
+				: new RelaxedDtlsCorrelationContextMatcher();
```

Comment:
For "backwards compatiblitiy". Mainly for some unit tests, which assume such dtls ... matcher. Reaching 8xx lines, I didn't wan't to change more with this PR.